### PR TITLE
Enable nullability in StronglyTypedResourceBuilder

### PR DESCRIPTION
**Pull request by halgab**
I had to fix nullable annotations of `ResXDataNode` in the process. This class was annotated for .Net 8, so it's not too late to make the changes without too much hassle.
I also unearthed quite a few problems in the annotations of System.CodeDom that forced me to use the null forgiving operator more than I should normally have to.

## Proposed changes

- Enable nullability in `StronglyTypedResourceBuilder`

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9663)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs [hunk 1]
```diff
@@ -153,30 +151,29 @@ public static partial class StronglyTypedResourceBuilder
...
                 {
                     throw new ArgumentException(string.Format(SR.MismatchedResourceName, keyname, node.Name));
                 }
 
-                string typeName = node.GetValueTypeName((AssemblyName[])null);
-                Type type = Type.GetType(typeName);
-                string valueAsString = node.GetValue((AssemblyName[])null).ToString();
+                string typeName = node.GetValueTypeName((AssemblyName[]?)null)!;
+                Type? type = Type.GetType(typeName);
+                string? valueAsString = node.GetValue((AssemblyName[]?)null)!.ToString();
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:176]: Why is this declared as nullable? you already use `!` before converting it to `oString()`

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs [hunk 2]
```diff
@@ -576,14 +573,14 @@ private static void AddGeneratedCodeAttributeforMember(CodeTypeMember typeMember
         //
         // return resourceMan;
 
-        CodeFieldReferenceExpression resourceManagerField = new(null, ResourceManagerFieldName);
+        CodeFieldReferenceExpression resourceManagerField = new(null!, ResourceManagerFieldName);
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: this doesn't seem right. Did we annotate `CodeFieldReferenceExpression`  incorrectly?

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs [hunk 3]
```diff
@@ -576,14 +573,14 @@ private static void AddGeneratedCodeAttributeforMember(CodeTypeMember typeMember
...
         CodeMethodReferenceExpression objectEqualsMethod =
             new(new CodeTypeReferenceExpression(typeof(object)), "ReferenceEquals");
 
         CodeMethodInvokeExpression isResourceManagerNull = new(
             objectEqualsMethod,
             resourceManagerField,
-            new CodePrimitiveExpression(null));
+            new CodePrimitiveExpression(null!));
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:583]: same here.

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 1 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:176]: `ToString` outputs a `string?`

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 2 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: I think so, yes

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 3 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:583]: I also think this was incorrectly annotated

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 2 above)
Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: Can you open an issue there and link to this?

Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: I am hesitant to take this to 8.0. Keeping under 9.0 and see if we get more details from runtime on this.

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: Sure

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 3 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:583]: FYI, I see we're not the first ones who bumped into this: dotnet/runtime#78703

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 2 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: I've put up a PR (#9667) that just fixes the annotations of `ResXDataNode`. I think it should be merged in .Net 8

Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs:576]: As for the runtime issue, it was already there: dotnet/runtime#78036

**Comment by elachlan** [comment]
@halgab do you plan to tackle the problems in runtime?

**Comment by halgab** [comment]
The dotnet team started to discuss the issue again in dotnet/runtime#78036. I'm waiting to see where this discussion is going before offering my help to annotate System.CodeDom.

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 3 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs]: see https://github.com/dotnet/winforms/pull/9663#issuecomment-1678834362

File: src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs (hunk 2 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/Resources/Tools/StronglyTypedResourceBuilder.cs]: see https://github.com/dotnet/winforms/pull/9663#issuecomment-1678834362

**Review by dreddy-work — APPROVED** [review]

**Pull request merged**